src/gccrs/args.rs

```diff
@@ -0,0 +1,152 @@
+            _ => CrateType::Unknown,
+        }
+    }
+}
+
+/// Structure used to represent arguments passed to `gccrs`. Convert them from `rustc`
+/// arguments using [`GccrsArg::from_rustc_arg`]
+pub struct GccrsArgs {
+    source_files: Vec<String>,
+    crate_type: CrateType,
+    output_file: String,
+}
+
+impl GccrsArgs {
+    fn generate_parser() -> Options {
+        let mut options = Options::new();
+        options.optopt("", "crate-name", "Name of the crate to compile", "NAME");
+        options.optopt("", "edition", "Rust edition to use", "YEAR");
+        options.optopt("", "error-format", "Requested error format", "EXTENSION");
+        options.optopt(
+            "",
+            "out-dir",
+            "Directory in which to output generated files",
+            "DIR",
+        );
+        options.optopt("", "emit", "Requested output to emit", "KIND");
+        options.optopt("", "json", "JSON Rendering type", "RENDER");
+        options.optopt("", "json", "JSON Rendering type", "RENDER");
+        options.optmulti("C", "", "Extra compiler options", "OPTION[=VALUE]");
+        options.optmulti(
+            "L",
+            "",
+            "Add a directory to the library's search path",
+            "KIND[=PATH]",
+        );
+        options.optmulti("", "crate-type", "Type of binary to output", "TYPE");
+
+        options
+    }
+
+    fn format_output_filename(matches: &Matches, crate_type: &CrateType) -> String {
+        // FIXME: No unwraps here
+        let crate_name = matches.opt_str("crate-name").unwrap();
+        let out_dir = matches.opt_str("out-dir").unwrap();
+
+        // FIXME: Figure out a way to return multiple output filenames. Just handle `bin`
+        // for now
+        let c_options = matches.opt_strs("C");
+
+        // FIXME: This is probably different on Windows, we should use Paths and PathStrs
+        // instead
+        let mut output_file = match crate_type {
+            &CrateType::Bin => format!("{}/{}", out_dir, crate_name),
+            &CrateType::DyLib => format!("{}/lib{}.so", out_dir, crate_name),
+            &CrateType::StaticLib => format!("{}/lib{}.a", out_dir, crate_name),
+            _ => unreachable!(
+                "gccrs cannot handle other crate types than bin, dylib or staticlib at the moment"
+            ),
+        };
+
+        // FIXME: Horrendous. We need to create a separate "C options" parser since we'll
+        // probably use more than just `extra-filename`.
+        c_options
+            .iter()
+            .for_each(|c_opt| match c_opt.split('=').nth(0).as_deref() {
+                Some("extra-filename") => output_file.push_str(&c_opt.split('=').nth(1).unwrap()),
+                _ => {}
+            });
+
+        dbg!(output_file)
+    }
+
+    /// Get the corresponding `gccrs` argument from a given `rustc` argument
+    pub fn from_rustc_args(rustc_args: &Vec<String>) -> Vec<GccrsArgs> {
+        let options = GccrsArgs::generate_parser();
```

Comment:
Why not use methods here instead of associated functions? That way you could do something like

```rust
GccrsArgs::new().from_rustc_args(args);
```

Or even better, implement it with the `TryFrom` trait, so you can do

```rs
GccrsArgs::try_from(args).unwrap(); // <- insert sane error handling here.
```

Reply:
The only issue with implementing `TryFrom` is that right now we need to return multiple sets of arguments for a single set of rust arguments. As mentionned in #3, `rustc` can generate mutiple types of binaries with a single command while gcc[rs] cannot (for now at least).

Reply:
Ah, I assumed from the name ``GccrsArgs`` that it already contained all the arguments and didn't see that this function returns a `Vec`.